### PR TITLE
Add --json and --silent options

### PR DESCRIPTION
**Pull request by kevinbarabash**
**Summary**
This PR adds `--json` and `--silent` options.  The `--json` option provides a way to get structured output that's easy to parse by other tools, e.g. and eslint rule I'm going to be working on.  The `--silent` option hides all output except for the JSON output.  This makes it easier for other tooling to consume the output.

**Details**
I facilitated the collection of data necessary for the JSON output by introducing the concept of middleware.  `FileReferenceLogger` uses the `JsonMiddleware` to collects an array of objects corresponding to the violations and errors that it logs.  Later on, process-cache.js logs the result of the middleware's `.output()` method which coverts the array to a JSON string.

**Issue** #634

**Review notes**
This is a draft PR at the moment.  There's still a number of tests that are failing and I wanted to get feedback on the overall design before working on fixing the tests.

### CODE REVIEW

File: src/process-cache.js [hunk 1]
```diff
@@ -19,9 +21,11 @@ export default async function processCache(
     const {autoFix} = options;
     const violationFileNames: Array<string> = [];
     const fileValidator = autoFix ? validateAndFix : validateAndReport;
+    const middlewares = options.json ? [new JsonMiddleware()] : [];
     for (const file of Object.keys(cache)) {
+        const fileRefLogger = new FileReferenceLogger(file, log, middlewares);
         try {
-            if (!(await fileValidator(options, file, cache, log))) {
+            if (!(await fileValidator(options, file, cache, fileRefLogger))) {
```
Comment (somewhatabstract) [inline src/process-cache.js:28]: You shouldn't need this.

File: src/validate-and-fix.js [hunk 2]
```diff
@@ -69,22 +75,22 @@ const reportBrokenEdge = (
...
-        ),
+    log.violation(
+        `${sourceFileRef} Updating checksum for sync-tag '${markerID}' referencing '${cwdRelativePath(
+            targetFile,
+        )}:${targetLine}' from ${
+            sourceChecksum || NO_CHECKSUM
+        } to ${targetChecksum}.`,
+        targetLine,
+        fix,
     );
```
Comment (somewhatabstract) [inline src/validate-and-fix.js:86]: I don't think the logging is the right place to change this stuff.

File: src/validate-and-fix.js [hunk 3]
```diff
@@ -154,13 +160,14 @@ const validateAndFix: FileProcessor = (
...
+                const fix = `${mappedFix == null ? line : mappedFix.fix}\n`;
                 if (mappedFix != null) {
-                    reportBrokenEdge(file, mappedFix.edge, log);
+                    reportBrokenEdge(file, mappedFix.edge, log, fix);
                 }
 
                 // If we have a fix, use it, otherwise, just output the line
                 // as it is (we have to add the newline).
-                ws.write(`${mappedFix == null ? line : mappedFix.fix}\n`);
+                ws.write(fix);
```
Comment (somewhatabstract) [inline src/validate-and-fix.js:170]: Why is this change required?

**Review by somewhatabstract — CHANGES_REQUESTED** [review]
OK, so I see how you've implemented this. However, I think it's in the wrong place and as such has overcomplicated the solution.

Instead, I think that a new validator should be made that is used in place of `validateAndFix`.

One that is `validateAndOutputJson` or something like that.

There shouldn't be a need to modify the logging, I don't think, because with a new validator, it will have control over the output. The only place that may need to be "silenced" in that mode are any messages output outside of the validator, which I think can just be turned off for the json mode.

**Review by somewhatabstract — COMMENTED** [review]
So, I think the places to change are:

1. A new validator here. for the json option
https://github.com/somewhatabstract/checksync/blob/bc65f9a57adfb8c798f7e1e607ce6ff919c462e3/src/process-cache.js#L21
2. Some changes in here to suppress the additional output when in json mode
https://github.com/somewhatabstract/checksync/blob/bc65f9a57adfb8c798f7e1e607ce6ff919c462e3/src/process-cache.js#L79-L90 
3. Here to suppress the extra log when in json mode
https://github.com/somewhatabstract/checksync/blob/bc65f9a57adfb8c798f7e1e607ce6ff919c462e3/src/process-cache.js#L94

Errors should still be logged and output to stderr. Anything listening for the stdout would only see the json and the exit code, regardless of the stderr log output.

I think the difficulty will be building the structured output. Right now the architecture is built to process file info and log violations one by one. To support JSON output, I think we should have a JSON builder that is created processing the cache - and it can have a `addFix` option, or similar, that allows a validator to capture a files violations. Then the validator can be curried around that, so there'd be a `makeJsonValidator(jsonBuilder)` call to make the validator function itself. The validator can then call that to add file violation info and fixes.

Then the main app can output that to stdout after the whole process is done when run with the `--json` mode. That way, we keep the json bits nice and separate from the other pieces.

**Comment by kevinbarabash** [comment]
@somewhatabstract creating a new validator sounds good.  I do have a question about how to deal with errors such as a sync-tag with a filename that doesn't exist.  Those errors aren't reported by the validators.  What would be a good way to go about capturing those for inclusion in the JSON output?

**Comment by somewhatabstract** [comment]
@kevinbarabash I expanded my earlier comment with some more architectural details around getting the JSON output.

> how to deal with errors such as a sync-tag with a filename that doesn't exist. Those errors aren't reported by the validators. What would be a good way to go about capturing those for inclusion in the JSON output?

So you are referring to this, I believe:
https://github.com/somewhatabstract/checksync/blob/bc65f9a57adfb8c798f7e1e607ce6ff919c462e3/src/generate-marker-edges.js#L130-L136

Yeah, I think that's a flaw in my design. File-specific errors should have always come from validators I think. To that end, we should generate an edge with a null target in that instance and then update the validators to properly handle that and output the error. That way validators cope with all file-specific errors. I think that would work and is certainly a better architecture.

**Comment by kevinbarabash** [comment]
I'll create a new PR with a new validator.
